File: dart_cli/lib/structs/JsonObject.dart

import "../ByteDataWrapper.dart";
import "../DataType.dart";
import "../KeysLookup.dart";
import "../SizeClass.dart";
import "../ioClass.dart";
import "Element.dart";

const _sizeClassToType = {
  SizeClass.empty: DataTypeChar.emptyObject,
  SizeClass.small: DataTypeChar.smallObject,
  SizeClass.big: DataTypeChar.bigObject,
  SizeClass.long: DataTypeChar.longObject
};
const _typeToSizeClass = {
  DataTypeChar.emptyObject: SizeClass.empty,
  DataTypeChar.smallObject: SizeClass.small,
  DataTypeChar.bigObject: SizeClass.big,
  DataTypeChar.longObject: SizeClass.long
};

class JsonObject extends JsonIoClass {
  final int length;
  final List<JsonObjectEntry> entries;
  final SizeClass sizeClass;

  JsonObject(this.length, this.entries, this.sizeClass);

  static (DataType, JsonObject?) fromJson(Object? json, KeysLookup keysLookup) {
    if (json is! Map)
      throw Exception("JsonObject.fromJson: json is not a Map: $json");
    var object = json;
    var length = object.length;
    var sizeClass = SizeClass.fromSize(length);
    var type = _sizeClassToType[sizeClass];
    if (type == null)
      throw Exception("JsonObject.fromJson: invalid size class: $sizeClass");
    var objectEntries = object.entries
      .map((e) => JsonObjectEntry.fromMapEntry(e as MapEntry<String, Object?>, keysLookup))
      .toList();
    return (
      DataType(type),
      sizeClass != SizeClass.empty
        ? JsonObject(length, objectEntries, sizeClass)
        : null
    );
  }

  static JsonObject readBytes(ByteDataWrapper bytes, DataType dataType, KeysLookup keysLookup) {
    var sizeClass = _typeToSizeClass[dataType.char];
    if (sizeClass == null)
      throw Exception("JsonObject.readBytes: dataType is not a object: $dataType");
    var length = sizeClass.readInt(bytes);
    var objectEntries = List.generate(
      length,
      (i) => JsonObjectEntry.readBytes(bytes, keysLookup)
    );
    return JsonObject(length, objectEntries, sizeClass);
  }

  @override
  Object? toJson() {
    return Map.fromEntries(entries.map((e) => e.toJson()));
  }

  @override
  void writeBytes(ByteDataWrapper bytes, KeysLookup keysLookup) {
    sizeClass.writeInt(bytes, length);
    for (var entry in entries)
      entry.writeBytes(bytes, keysLookup);
  }

  @override
  int get size {
    var size = sizeClass.bytes;
    for (var entry in entries)
      size += entry.size;
    return size;
  }
}

class JsonObjectEntry extends JsonIoClass {
  final String key;
  final Element value;

  JsonObjectEntry(this.key, this.value);

  static JsonObjectEntry fromMapEntry(MapEntry<String, Object?> entry, KeysLookup keysLookup) {
    var key = entry.key;
    var value = Element.fromJson(entry.value, keysLookup);
    keysLookup.visitKey(key);
    return JsonObjectEntry(key, value);
  }

  static JsonObjectEntry readBytes(ByteDataWrapper bytes, KeysLookup keysLookup) {
    var key = keysLookup.lookupIndex(bytes.readUint16());
    var value = Element.readBytes(bytes, keysLookup);
    return JsonObjectEntry(key, value);
  }

  @override
  MapEntry toJson() {
    return MapEntry(key, value.toJson());
  }

  @override
  void writeBytes(ByteDataWrapper bytes, KeysLookup keysLookup) {
    bytes.writeUint16(keysLookup.getKeyIndex(key));
    value.writeBytes(bytes, keysLookup);
  }

  @override
  int get size {
    return 2 + value.size;
  }
}

File: python_lib/structs/JsonObject.py
from __future__ import annotations
from DataType import DataType, DataTypeChar
from FileWrapper import BinaryReader, BinaryWriter
from KeysLookup import KeysLookup
from SizeClass import SizeClass
from ioClass import JsonIoClass
import structs.Element as Element


_sizeClassToType = {
	SizeClass.empty: DataTypeChar.emptyObject,
	SizeClass.small: DataTypeChar.smallObject,
	SizeClass.big: DataTypeChar.bigObject,
	SizeClass.long: DataTypeChar.longObject
}
_typeToSizeClass = {
	DataTypeChar.emptyObject: SizeClass.empty,
	DataTypeChar.smallObject: SizeClass.small,
	DataTypeChar.bigObject: SizeClass.big,
	DataTypeChar.longObject: SizeClass.long
}

class JsonObject(JsonIoClass):
	length: int
	entries: list[JsonObjectEntry]
	sizeClass: SizeClass

	def __init__(self, length: int, entries: list[JsonObjectEntry], sizeClass: SizeClass):
		self.length = length
		self.entries = entries
		self.sizeClass = sizeClass

	@staticmethod
	def fromJson(json: object, keysLookup: KeysLookup) -> tuple[DataType, JsonObject|None]:
		if not isinstance(json, dict):
			raise Exception(f"JsonObject.fromJson: json is not a dict: {json}")
		object = json
		length = len(object)
		sizeClass = SizeClass.fromSize(length)
		type = _sizeClassToType[sizeClass]
		if type is None:
			raise Exception(f"JsonObject.fromJson: invalid size class: {sizeClass}")
		entries = [
			JsonObjectEntry.fromMapEntry(entry, keysLookup)
			for entry in object.items()
		]
		return (
			DataType(type),
			JsonObject(length, entries, sizeClass) if sizeClass != SizeClass.empty else None
		)
	
	@staticmethod
	def readBytes(bytes: BinaryReader, dataType: DataType, keysLookup: KeysLookup) -> JsonObject:
		sizeClass = _typeToSizeClass[dataType.char]
		if sizeClass is None:
			raise Exception(f"JsonObject.readBytes: dataType is not a object: {dataType}")
		length = sizeClass.readInt(bytes)
		entries = [
			JsonObjectEntry.readBytes(bytes, keysLookup)
			for _ in range(length)
		]
		return JsonObject(length, entries, sizeClass)
	
	def toJson(self) -> object:
		return {entry.key: entry.value.toJson() for entry in self.entries}
	
	def writeBytes(self, bytes: BinaryWriter, keysLookup: KeysLookup) -> None:
		self.sizeClass.writeInt(bytes, self.length)
		for entry in self.entries:
			entry.writeBytes(bytes, keysLookup)

	@property
	def size(self) -> int:
		size = self.sizeClass.bytes
		for entry in self.entries:
			size += entry.size
		return size
	
class JsonObjectEntry(JsonIoClass):
	key: str
	value: Element.Element

	def __init__(self, key: str, value: Element.Element):
		self.key = key
		self.value = value

	@staticmethod
	def fromMapEntry(entry: tuple[str, object], keysLookup: KeysLookup) -> JsonObjectEntry:
		key = entry[0]
		value = Element.Element.fromJson(entry[1], keysLookup)
		keysLookup.visitKey(key)
		return JsonObjectEntry(key, value)
	
	@staticmethod
	def readBytes(bytes: BinaryReader, keysLookup: KeysLookup) -> JsonObjectEntry:
		key = keysLookup.lookupIndex(bytes.readUint16())
		value = Element.Element.readBytes(bytes, keysLookup)
		return JsonObjectEntry(key, value)
	
	def toJson(self) -> tuple[str, object]:
		return (self.key, self.value.toJson())
	
	def writeBytes(self, bytes: BinaryWriter, keysLookup: KeysLookup) -> None:
		bytes.writeUint16(keysLookup.getKeyIndex(self.key))
		self.value.writeBytes(bytes, keysLookup)

	@property
	def size(self) -> int:
		return 2 + self.value.size
